system/jlib/jdebug.hpp

```diff
@@ -225,6 +226,107 @@ class jlib_decl HiresTimer
 };
 
 
+//---------------------------------------------------------------------------------------------------------------------
+
+/*
+There are several situations where we want to record the time spent waiting for items to be processed - time an item
+is queued before processing, time spent calling LDAP from within esp etc.. Gathering the metric has a few complications:
+* The metric gathering needs to have minimal impact
+* The waiting time needs to include items that are currently waiting as well completed
+* Multiple items can be being waited for at the same time, and complete at different times.
+
+    waitingTime = sum(itemTime)
+    = sum(completedItemTime) + sum(inflightItemTime)
+    = sum(endCompletedTime-startCompletedTime) + sum(currentTime-startInflightTime))
+    = sumEndCompletedTime - sumStartCompletedTime + sum(currentTime) - sumStartInflightTime
+    = sumEndCompletedTime + num*current - sumStartTime
+
+The following pattern efficiently solves the problem:
+
+ * When an item is queued/started atomically increment a counter, and add the current timestamp to the accumulated start time stamps.
+ * When an items is dequeued/complete atomically decrement a counter and add the current timestamp to the accumulated end time stamps.
+ * To calculate the waiting time use (sumEndTimestamps + numWaiting * currentTimeStamp) - sumStartTimestamps.
+
+
+At first glance this appears to have a problem because sumEndTimestamps will quickly overflow, but because of the
+properties of modulo arithmetic, the final result will be correct even if it has overflowed!
+Also since you are only ever interested in (sumEndTimestamps - sumStartTimestamps) the same field can be used for both.
+
+There are two versions, one that uses a critical section, and a second that uses atomics, but is limited to the number
+of active blocked items.
+*/
+
+class jlib_decl BlockedTimeTracker
+{
+public:
+    BlockedTimeTracker() = default;
+    BlockedTimeTracker(const BlockedTimeTracker &) = delete;
+
+    void noteWaiting();
+    void noteComplete();
+    __uint64 getWaitingNs() const;
+
+private:
+    mutable CriticalSection cs;
+    unsigned numWaiting = 0;
+    cycle_t timeStampTally = 0;
+};
+
+class jlib_decl BlockedSection
+{
+public:
+    BlockedSection(BlockedTimeTracker & _tracker) : tracker(_tracker) { tracker.noteWaiting(); }
+    ~BlockedSection() { tracker.noteComplete(); }
+private:
+    BlockedTimeTracker & tracker;
+};
+
+//Lightweight version that uses a single atomic, but has a limit on the number of active blocked items
+//Easier to understand by looking at the code for the BlockedTimeTracker class
+//Could be a template class (for MAX_ACTIVE), but I doubt it is worth it.
+class jlib_decl LightweightBlockedTimeTracker
+{
+    static constexpr unsigned MAX_ACTIVE = 64;  //MAX_ACTIVE should be a power of 2 for efficiency
```

Comment:
At 2^32 cycles/sec (approx) you can measure about 2^24 seconds before you start to get issues with loss of resolution, I think... 

If so MAX_ACTIVE could safely be a bit higher, I suspect.

Reply:
I think it currently gives about 2 years accumulated blocked time.  Hard to know at the moment what a sensible value would be.